restful_api/myapi/api/resources/quiz_attempt.py

```diff
@@ -67,21 +72,142 @@ def get(self, course_id, section_id, trainer_id, eng_id):
+                try:
+                    db.session.add(quiz_attempt)
+                    db.session.add(section_completed)
+                    db.session.commit()
+                except IntegrityError as e:
+                    return {"msg": str(e)}, 400
+
+                return self.return_output_passed(self, base_output, section_completed, enrollment)
+
+            else:
+                return self.return_output_failed(self, base_output)
+
+    @staticmethod
+    def return_output_completed(self, base_output, section_completed):
+        return {
+            "msg": "quiz attempt created",
+            "grade": "completed",
+            "score": base_output[0],
+            "quiz_attempt": self.schema.dump(base_output[2]),
+            "wrong_answers": base_output[1],
+            "section_completed": self.section_completed_schema.dump(section_completed),
+        }, 201
+
+    @staticmethod
+    def return_output_passed(self, base_output, section_completed, enrollment):
+        return {
+            "msg": "quiz attempt created",
+            "grade": "passed",
+            "score": base_output[0],
+            "quiz_attempt": self.schema.dump(base_output[2]),
+            "wrong_answers": base_output[1],
+            "section_completed": self.section_completed_schema.dump(section_completed),
+            "updated_enrollment": self.enroll_schema.dump(enrollment)
+        }, 201
+
+    @staticmethod
+    def return_output_failed(self, base_output):
+        return {
+            "msg": "quiz attempt created",
+            "grade": "failed",
+            "score": base_output[0],
+            "quiz_attempt": self.schema.dump(base_output[2]),
+            "wrong_answers": base_output[1]
+        }, 201
+
+    @staticmethod
+    def grade_quiz(quiz, answers):
+        questions = quiz['questions']
+        # Format to dict {question_id : correct_option_label}
+        fmted_questions = {}
+        for question in questions:
+            question_id = question['question_id']
+            question_options = question['question_options']
+            for question_option in question_options:
+                if question_option['is_correct'] == True:
+                    correct_option_label = question_option['option_label']
+                    fmted_questions[question_id] = correct_option_label
+
+        # Sort fmted_questions and answers by question_id
+        fmted_questions = sorted(fmted_questions.items(), key=lambda d: d[0])
+        answers = sorted(answers, key=lambda d: d['question_id'])
+
+        # If quiz_attempt incomplete, set answer_labels for unattempted questions to be None
+        if len(answers) < len(fmted_questions):
+            for i in range(len(fmted_questions) - len(answers)):
+                answers.append({"answer_label": None})
+
+        wrong = []
+        num_correct = 0
+        for i in range(len(fmted_questions)):
+            if fmted_questions[i][1] == answers[i]['answer_label']:
```

Comment:
May at the start of for loop just add a sanity check to make sure the question and answers are the same.

```
assert fmted_questions[i][0] == answers[i]['question_id'], "Wrong question and answer matching"

```

Reply:
ok!